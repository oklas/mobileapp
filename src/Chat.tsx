import React from 'react';

import { GiftedChat } from 'react-native-gifted-chat';
import Connection from 'wrtc-simple-data-native';

/*
  <Chat
    wrtcDataParams={{
      ...config // onfigure `wrtc-simple-data` connection
    }}
    logger={(msg: string) => console.log(msg)}
    messages=[] // initial messages if any
  />
*/

export type WrtcDataConnectionOptions = {
  signallingServer ?: string
  roomName ?: string
  rtcOpts ?: Object
  channelName ?: string
  channelOpts ?: Object
  debugMode ?: boolean
  debugLogger ?: (msg: string) => void
}

export interface Props {
  wrtcDataParams ?: WrtcDataConnectionOptions
  messages ?: Object[]
  logger ?: (msg: string) => void
}

export interface State {
  messages: Object[],
  ready: boolean,
}

export default class Chat extends React.Component<Props, State> {
  state: State
  connection: Connection
  users: Object
  newMessageId: number
  wrtcDataParams: WrtcDataConnectionOptions
  logger: (msg: string) => void

  constructor(props: Props) {
    super(props)
    this.state = {
      messages: props.messages || [],
      ready: false
    }

    this.users = {
      'local': {
        _id: 'local',
        name: 'Local User',
      },
      'system': {
        _id: 'system',
        name: '',
      }
    }

    this.newMessageId = 1

    this.logger = props.logger ? props.logger : (msg: string) => {
      this.appendMessage('system', msg)
    }

    let wdp: WrtcDataConnectionOptions = props.wrtcDataParams || {}

    this.wrtcDataParams = {
      roomName: wdp.roomName || 'chatRoom',
      signallingServer: wdp.signallingServer || 'http://localhost:3000',
      rtcOpts: wdp.rtcOpts || {iceServers: [{urls: 'stun:stun.l.google.com:19301'}]},
      debugMode: wdp.debugMode || false,
      debugLogger: wdp.debugLogger || this.logger,
    }
  }

  findUser = (userId) => {
    let user = this.users[userId]
    if( !user ) {
      user = {
        _id: userId,
        name: userId,
      }
      this.users[userId] = user
    }
    return user
  }

  appendMessage = (userId: string, message: string) => {
    const messages = [{
      _id: this.newMessageId++,
      text: message,
      createdAt: new Date(),
      user: this.findUser(userId),
    }]
    this.setState((previousState) => ({
      messages: GiftedChat.append(previousState.messages, messages),
    }))
  }

  onSend = (messages = []) => {
    if(this.state.ready) {
      this.setState((previousState) => ({
        messages: GiftedChat.append(previousState.messages, messages),
      }))
      messages.forEach(item => this.connection.sendMessage(item.text))
    }
  }

  componentDidMount() {
    this.connection = new Connection(this.wrtcDataParams)

    this.connection.on('ready', () => {
      this.logger('connected to the server')
      this.connection.on('channel:ready', () => {
        this.logger('new peer connected')
        this.setState({ready: true})
      })
      this.connection.on('message', (data) => {
        this.appendMessage(data.sender, data.text)
      })
    })
  }

  render() {
    return (
      <GiftedChat
        messages={this.state.messages}
        onSend={this.onSend}
        user={this.findUser('local')}
      />
    )
  }
}
