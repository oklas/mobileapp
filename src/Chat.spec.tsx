import 'react-native'
import React from 'react'
import Chat from './Chat'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';


jest.mock('wrtc-simple-data-native', () => {
  return require('wrtc-simple-data')
});

jest.mock('react-native-gifted-chat', () => {
  return require('./test/GiftedChat.mock');
});

// https://github.com/airbnb/enzyme/issues/614

it('renders correctly', () => {
  const messages = [{
    text: 'test message',
    createdAt: new Date(),
    user: {
      _id: 'user',
      name: 'User',
    }
  }]

  const tree = renderer.create(
    <Chat
      wrtcDataParams={{debugMode: true}}
      messages={messages}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
  expect(tree).toBeDefined();
});



