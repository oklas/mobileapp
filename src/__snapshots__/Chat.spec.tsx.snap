// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      
    </Text>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        debug enabled
      </Text>
    </View>
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      User
    </Text>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        test message
      </Text>
    </View>
  </View>
</View>
`;
